k8s-operator/sessionrecording/tsrecorder/tsrecorder.go

```diff
@@ -44,9 +45,11 @@ type Client struct {
 	conn io.WriteCloser // connection to a tsrecorder instance
 }
 
-// Write appends timestamp to the provided bytes and sends them to the
-// configured tsrecorder.
-func (rec *Client) Write(p []byte) (err error) {
+// WriteOutput can be used to send data output to the terminal to the tsrecorder.
+// This can be data sent to STDOUT or STDERR streams.
+// https://docs.asciinema.org/manual/asciicast/v2/#o-output-data-written-to-a-terminal
+func (rec *Client) WriteOutput(p []byte) (err error) {
+	const outputEventCode = "o"
 	if len(p) == 0 {
 		return nil
 	}
```

Comment:
> Same comment for the length check above I guess as well?

We can't remove the length check (since this code adds timestamp and event code, the bytes would no longer be empty when received in `write`)

> Performance benefits from early return seem negligible

I've not done performance tests on json, but it does seem wasteful to be marshalling each written message for a session that is no longer recorded and it does not seem that this check adds too much complexity here? I've removed the backoff and length checks from `write` instead

Reply:
... ended up pulling the json marshalling down so, the check for failed recording state can now be done in a single place https://github.com/tailscale/tailscale/pull/12965/commits/77a03cc7b52c17beb1d266d1e525c6aca140717e#diff-3f43ab0d05f05c448214ad76dbbefd94b56a64d518f8a8f89ab97a1a2ed56fb0R85-R90